src/application/machines/devtoolsMachine.tsx

```diff
@@ -1,143 +1,233 @@
+
+export type EmittedEvents =
+  Extract<Events, { type: `emit.${string}` }> extends infer EmitTriggerEvents
+    ? {
+        [K in keyof EmitTriggerEvents]: K extends "type"
+          ? EmitTriggerEvents[K] extends `emit.${infer Type}`
+            ? Type
+            : EmitTriggerEvents[K]
+          : EmitTriggerEvents[K];
+      }
+    : never;
 
 export const devtoolsMachine = setup({
-  types: {
-    context: {} as { modalOpen: boolean },
-    events: {} as Events,
+  types: {} as {
+    context: DevtoolsMachineContext;
+    events: Events;
+    children: { reconnect: "reconnect" };
+    emitted: EmittedEvents;
   },
   delays: {
     connectTimeout: 10_000,
   },
   actions: {
-    openModal: assign({ modalOpen: true }),
-    closeModal: assign({ modalOpen: false }),
-    notifyWaitingForConnection: () => {
-      BannerAlert.show({
-        type: "loading",
-        content: "Waiting for client to connect...",
-      });
-    },
     closeBanner: BannerAlert.close,
     notifyDisconnected: () => {
       BannerAlert.show({
         type: "loading",
-        content: "Disconnected. Waiting for client to connect...",
+        content: "No client connected. Waiting for client to connect...",
       });
     },
     notifyConnected: () => {
       BannerAlert.show({ type: "success", content: "Connected!" });
     },
-    notifyTimedOut: () => {
-      BannerAlert.show({
-        type: "error",
-        content:
-          "Unable to communicate with browser tab. Please reload the window and restart the devtools to try again.",
-      });
-    },
-    notifyNotFound: ({ self }) => {
-      BannerAlert.show({
-        type: "error",
-        content: (
-          <div className="flex justify-between items-center">
-            Client not found{" "}
-            <Button
-              size="xs"
-              variant="hidden"
-              icon={<IconSync />}
-              onClick={() => self.send({ type: "retry" })}
-            >
-              Retry connection
-            </Button>
-          </div>
-        ),
-      });
+
+    renderUI: () => {
+      throw new Error("Provide implementation");
     },
     resetStore: () => {
-      throw new Error("Provide implementation in the component");
+      throw new Error("Provide implementation");
     },
   },
+  guards: {
+    contextValid: ({ context }) => {
+      return context.listening !== false;
+    },
+  },
+  actors: {
+    reconnect: reconnectMachine,
+  },
 }).createMachine({
   id: "devtools",
-  initial: "initialized",
+  type: "parallel",
   context: {
-    modalOpen: false,
+    modals: {},
+    port: undefined,
+    registeredClients: 0,
   },
   on: {
-    closeModal: {
-      actions: "closeModal",
+    // forward reconnect events to child actor
+    "reconnect.*": {
+      actions: sendTo("reconnect", ({ event }) => event),
+    },
+    // on an `emit.*` event, `emit` that event so it can be subscribed to from the app
+    "emit.*": {
```

Comment:
Before https://github.com/apollographql/apollo-client-devtools/pull/1536/files, this would allow us to call `refetch` as a reaction to that event.

It's not like we need this anymore *right now*, but it is a good blueprint we can build on if we need something like this in the future, and it doesn't hurt, so I'd keep it in for now.